nexus/deployment/src/blueprint_builder.rs

```diff
@@ -0,0 +1,690 @@
+                            .get(sled_id)
+                            .cloned()
+                    })
+                    // If it's not there either, then this must be a new sled
+                    // and we haven't added any zones to it yet.  Use the
+                    // standard initial config.
+                    .unwrap_or_else(|| OmicronZonesConfig {
+                        generation: Generation::new(),
+                        zones: vec![],
+                    });
+                (*sled_id, zones)
+            })
+            .collect();
+        Blueprint {
+            id: Uuid::new_v4(),
+            omicron_zones: omicron_zones,
+            zones_in_service: self.zones_in_service,
+            parent_blueprint_id: Some(self.parent_blueprint.id),
+            time_created: chrono::Utc::now(),
+            creator: self.creator,
+            comment: self.comments.join(", "),
+        }
+    }
+
+    /// Sets the blueprints "comment"
+    ///
+    /// This is a short human-readable string summarizing the changes reflected
+    /// in the blueprint.  This is only intended for debugging.
+    pub fn comment<S>(&mut self, comment: S)
+    where
+        String: From<S>,
+    {
+        self.comments.push(String::from(comment));
+    }
+
+    pub fn sled_ensure_zone_ntp(
+        &mut self,
+        sled_id: Uuid,
+    ) -> Result<Ensure, Error> {
+        // If there's already an NTP zone on this sled, do nothing.
+        let has_ntp = self
+            .parent_blueprint
+            .omicron_zones
+            .get(&sled_id)
+            .map(|found_zones| {
+                found_zones.zones.iter().any(|z| z.zone_type.is_ntp())
+            })
+            .unwrap_or(false);
+        if has_ntp {
+            return Ok(Ensure::NotNeeded);
+        }
+
+        let sled_info = self.sled_resources(sled_id)?;
+        let sled_subnet = sled_info.subnet;
+        let ip = self.sled_alloc_ip(sled_id)?;
+        let ntp_address = SocketAddrV6::new(ip, NTP_PORT, 0, 0);
+
+        // Construct the list of internal DNS servers.
+        //
+        // It'd be tempting to get this list from the other internal NTP
+        // servers but there may not be any of those.  We could also
+        // construct this list manually from the set of internal DNS servers
+        // actually deployed.  Instead, we take the same approach as RSS:
+        // these are at known, fixed addresses relative to the AZ subnet
+        // (which itself is a known-prefix parent subnet of the sled subnet).
+        let dns_servers =
+            get_internal_dns_server_addresses(sled_subnet.net().network());
+
+        // The list of boundary NTP servers is not necessarily stored
+        // anywhere (unless there happens to be another internal NTP zone
+        // lying around).  Recompute it based on what boundary servers
+        // currently exist.
+        let ntp_servers = self
+            .parent_blueprint
```

Comment:
This, like other parts of the builder that rely on the `parent_blueprint` seem to implicitly assume that the `parent_blueprint` was successfully executed. That is why they are capable of performing "next steps". Am I correct here?

If I am correct then this may present a problem when a plan is not successfully executed, or partially executed. While conceptually I like the idea of one plan building off the next plan, it seems like we may need to pass in more parameters where this may be problematic, or we need to use the inventory rather than the prior plan.

Reply:
I wondered this too, but wasn't sure how to phrase the question, because I assume it's pretty reasonable we end up in this case:

* We have a current target blueprint that we have not yet achieved
* Something in the system changes such that we can never achieve the current target

Now what? We either need to be able to generate a child blueprint despite the parent not being successfully executed, or maybe we need some escape hatch to say "regenerate a new initial blueprint based on the latest collection"?

Reply:
> I wondered this too, but wasn't sure how to phrase the question, because I assume it's pretty reasonable we end up in this case:
> 
>     * We have a current target blueprint that we have not yet achieved
> 
>     * Something in the system changes such that we can never achieve the current target
> 
> 
> Now what? We either need to be able to generate a child blueprint despite the parent not being successfully executed, or maybe we need some escape hatch to say "regenerate a new initial blueprint based on the latest collection"?

Agreed.  I think the conclusion is that we can't assume the parent blueprint has/will be executed.  So far, I don't think that's been a difficult constraint.

Reply:
Your third principle covers what I was worried about. Specifically if we started to deploy crucible zones assuming that the internal NTP zone was deployed already. If the planner takes into consideration the current inventory and that the parent blueprint may not have executed before generating the next blueprint, then we should be all set. This also means that the planner may very well be in a "wait for the target blueprint to complete" before it does anything for a good amount of time. That seems normal though.